source/common/upstream/cluster_manager_impl.cc

```diff
@@ -988,6 +993,29 @@ void ClusterManagerImpl::ThreadLocalClusterManagerImpl::onHostHealthFailure(
   }
 }
 
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::ConnPoolsContainer&
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::getOrAllocateHttpConnPoolsContainer(
+    const HostConstSharedPtr& host) {
+  auto container_iter = host_http_conn_pool_map_.find(host);
+  if (container_iter == host_http_conn_pool_map_.end()) {
+    ConnPoolsContainer container{thread_local_dispatcher_};
+    container_iter = host_http_conn_pool_map_.emplace(host, std::move(container)).first;
+  }
+
+  return container_iter->second;
+}
+
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::ConnPoolsContainer*
+ClusterManagerImpl::ThreadLocalClusterManagerImpl::getHttpConnPoolsContainer(
```

Comment:
I would probably merge this and above into a single method with an `allocate` bool param, and just return a pointer, but I don't feel strongly about it. TIOLI.